### ISSUE TITLE
No response from remote cluster with Istio enabled

### ISSUE THREAD

**Issue by davtex** [cilium/cilium#34015]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

higher than v1.15.7 and lower than v1.16.0

### What happened?

Hi,

I am running two OKD (Openshift) 4.14 clusters with Cilium 1.15.1 (latest available for Openshift) in ClusterMesh and I am trying to get Istio Multicluster (Multi-Primary, single network) to work with utilizing ClusterMesh as alternative to previously used Submariner. We want to utilize Cilium for most of our workloads with Istio as an option for workloads that are already using Istio and some of its more advanced L7 features.

I have Istio sleep and helloworld applications running in "sample" namespace and the whole setup is mostly working. For testing I am using curl from cluster1 helloworld, calling helloworld service (expecting loadbalanced responses from cluster1 and cluster2) and also directly pod in cluster2.

Use case 1:
- cilium + no Istio + global service enabled
- calling helloworld.sample -> OK, I receive responses from both clusters
- calling podIP in cluster2 -> OK

Use case 2:
- cilium + Istio + global service disabled
- calling helloworld.sample -> NOK, I receive response from cluster1, but I get no response from cluster2
- calling podIP in cluster2 -> OK

TL;DR -> With Istio enabled and utilizing Clustermesh, I can reach pod in second cluster without issues, which tells me that both Cilium Clustermesh and Istio Multicluster are working properly. But when I try do reach the second cluster via ClusterIP service, I get no response.

To make it even more puzzling, even though I get no response in curl, I can clearly see in Istio sidecar logs that request was received and HTTP response was sent. It just never arrives back to cluster1. As you can see below, communication looks almost the same with difference that when I use ClusterIP service, I get no reply in curl and I see additional repeated ACK messages that continue until I terminate running curl.

Calling POD directly (curl 100.x.x.x:5000/hello):
```
SOURCE:
-> overlay flow 0x913a2107 , identity 719155->756032 state new ifindex cilium_vxlan orig-ip 0.0.0.0: 100.70.7.81:49010 -> 100.71.7.216:5000 tcp SYN
-> endpoint 4030 flow 0xad8936c7 , identity 756032->719155 state reply ifindex lxc358277867803 orig-ip 100.71.7.216: 100.71.7.216:5000 -> 100.70.7.81:49010 tcp SYN, ACK
-> overlay flow 0x913a2107 , identity 719155->756032 state established ifindex cilium_vxlan orig-ip 0.0.0.0: 100.70.7.81:49010 -> 100.71.7.216:5000 tcp ACK
-> overlay flow 0x913a2107 , identity 719155->756032 state established ifindex cilium_vxlan orig-ip 0.0.0.0: 100.70.7.81:49010 -> 100.71.7.216:5000 tcp ACK
-> endpoint 4030 flow 0xad8936c7 , identity 756032->719155 state reply ifindex lxc358277867803 orig-ip 100.71.7.216: 100.71.7.216:5000 -> 100.70.7.81:49010 tcp ACK


TARGET:
-> endpoint 2214 flow 0x90b499ce , identity 719155->756032 state new ifindex lxc73cdcd6456bd orig-ip 100.70.7.81: 100.70.7.81:49010 -> 100.71.7.216:5000 tcp SYN
-> overlay flow 0xf3a77fb3 , identity 756032->719155 state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 100.71.7.216:5000 -> 100.70.7.81:49010 tcp SYN, ACK
-> endpoint 2214 flow 0x90b499ce , identity 719155->756032 state established ifindex lxc73cdcd6456bd orig-ip 100.70.7.81: 100.70.7.81:49010 -> 100.71.7.216:5000 tcp ACK
-> endpoint 2214 flow 0x90b499ce , identity 719155->756032 state established ifindex lxc73cdcd6456bd orig-ip 100.70.7.81: 100.70.7.81:49010 -> 100.71.7.216:5000 tcp ACK
-> overlay flow 0xf3a77fb3 , identity 756032->719155 state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 100.71.7.216:5000 -> 100.70.7.81:49010 tcp ACK
```

Calling POD via service (curl helloworld.sample:5000/hello):
```
SOURCE:
-> overlay flow 0xcc29c186 , identity 719155->756032 state new ifindex cilium_vxlan orig-ip 0.0.0.0: 100.70.7.81:50296 -> 100.71.7.216:5000 tcp SYN
-> endpoint 4030 flow 0xab69702c , identity 756032->719155 state reply ifindex lxc358277867803 orig-ip 100.71.7.216: 100.71.7.216:5000 -> 100.70.7.81:50296 tcp SYN, ACK
-> overlay flow 0xcc29c186 , identity 719155->756032 state established ifindex cilium_vxlan orig-ip 0.0.0.0: 100.70.7.81:50296 -> 100.71.7.216:5000 tcp ACK
-> overlay flow 0xcc29c186 , identity 719155->756032 state established ifindex cilium_vxlan orig-ip 0.0.0.0: 100.70.7.81:50296 -> 100.71.7.216:5000 tcp ACK
-> endpoint 4030 flow 0xab69702c , identity 756032->719155 state reply ifindex lxc358277867803 orig-ip 100.71.7.216: 100.71.7.216:5000 -> 100.70.7.81:50296 tcp ACK


TARGET:
-> endpoint 2214 flow 0x9f1f5cb3 , identity 719155->756032 state new ifindex lxc73cdcd6456bd orig-ip 100.70.7.81: 100.70.7.81:50296 -> 100.71.7.216:5000 tcp SYN
-> overlay flow 0xfddbc55a , identity 756032->719155 state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 100.71.7.216:5000 -> 100.70.7.81:50296 tcp SYN, ACK
-> endpoint 2214 flow 0x9f1f5cb3 , identity 719155->756032 state established ifindex lxc73cdcd6456bd orig-ip 100.70.7.81: 100.70.7.81:50296 -> 100.71.7.216:5000 tcp ACK
-> endpoint 2214 flow 0x9f1f5cb3 , identity 719155->756032 state established ifindex lxc73cdcd6456bd orig-ip 100.70.7.81: 100.70.7.81:50296 -> 100.71.7.216:5000 tcp ACK
-> overlay flow 0xfddbc55a , identity 756032->719155 state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 100.71.7.216:5000 -> 100.70.7.81:50296 tcp ACK
-> overlay flow 0xcaee3d20 , identity 756032->719155 state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 100.71.7.216:5000 -> 100.70.7.81:50296 tcp ACK
-> overlay flow 0xc048f53b , identity 756032->719155 state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 100.71.7.216:5000 -> 100.70.7.81:50296 tcp ACK
-> overlay flow 0x180100d6 , identity 756032->719155 state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 100.71.7.216:5000 -> 100.70.7.81:50296 tcp ACK
```

I am pretty sure I am missing something obvious here that prevents this communication from working.

### How can we reproduce the issue?

1. Install 2 OKD 4.14 clusters in AWS

Cluster1 subnets:
Pod: 100.70.0.0/16
Svc: 172.30.0.0/16

Cluster2 subnets:
Pod: 100.71.0.0/16
Svc: 172.31.0.0/16

3. Configure CIlium with CiliumConfig below

Cluster1:
```
apiVersion: cilium.io/v1alpha1
kind: CiliumConfig
metadata:
  name: cilium
  namespace: cilium
spec:
  bpf:
    # 
    masquerade: true
  cluster:
    id: 10
    name: mer-dev01-mercury-dev01
  clusterHealthPort: 9940
  clustermesh:
    apiserver:
      replicas: 2
      service:
        annotations:
          service.beta.kubernetes.io/aws-load-balancer-internal: "true"
          service.beta.kubernetes.io/aws-load-balancer-type: nlb
          external-dns.alpha.kubernetes.io/hostname: cilium.mercury-dev01.mer-dev01.internal.ABC.cloud
        type: LoadBalancer
      tls:
        auto:
          enabled: false
        server:
          extraDnsNames:
            - cilium.mercury-dev01.mer-dev01.internal.ABC.cloud
          cert: "ABC"
          key: "ABC"
        # -- base64 encoded PEM values for the clustermesh-apiserver admin certificate and private key.
        # Used if 'auto' is not enabled.
        admin:
          cert: "ABC"
          key: "ABC"
        # -- base64 encoded PEM values for the clustermesh-apiserver remote cluster certificate and private key.
        # Used if 'auto' is not enabled.
        remote:
          cert: "ABC"
          key: "ABC"
        client:
          cert: "ABC"
          key: "ABC"
    config:
      clusters:
      - address: cilium.mercury-dev01.mer-dev02.internal.ABC.cloud
        name: mer-dev02-mercury-dev01
        port: 2379
      enabled: true
    useAPIServer: true
  cni:
    binPath: /var/lib/cni/bin
    chainingMode: portmap
    confPath: /var/run/multus/cni/net.d
    # # Needed for Istio
    exclusive: false
  debug:
    enabled: true
  etcd:
    leaseTTL: 30s
  externalWorkloads:
    enabled: true
  hubble: # Enable Hubble and Hubble Relay and avoid UI issues
    enabled: true
    peerService:
      clusterDomain: cluster.local
    relay:
      enabled: true
      tls:
        client:
          cert: "ABC"
          key: "ABC"
    tls:
      auto:
        enabled: false
      server:
        cert: "ABC"
        key: "ABC"
    ui:
      enabled: true
  identityChangeGracePeriod: 0s
  ipam:
    mode: cluster-pool
    operator:
      clusterPoolIPv4MaskSize: "23"
      clusterPoolIPv4PodCIDRList:
      - 100.70.0.0/16
  ipv4:
    enabled: true
  ipv4NativeRoutingCIDR: 100.70.0.0/15
  ipv6:
    enabled: false
  k8s:
    requireIPv4PodCIDR: true
  k8sServiceHost: api.mercury-dev01.mer-dev01.internal.ABC.cloud
  k8sServicePort: 6443
  kubeProxyReplacement: true
  logSystemLoad: true
  prometheus:
    serviceMonitor:
      enabled: false
  rollOutCiliumPods: true
  securityContext:
    privileged: true
  sessionAffinity: true
  tunnelPort: 4789
  tls:
    ca:
      cert: ABC
      key: ABC
  # Needed for Istio
  socketLB:
    hostNamespaceOnly: true
```

Cluster2:
```
apiVersion: cilium.io/v1alpha1
kind: CiliumConfig
metadata:
  name: cilium
  namespace: cilium
spec:
  bpf:
    masquerade: true
  cluster:
    id: 11
    name: mer-dev02-mercury-dev01
  clusterHealthPort: 9940
  clustermesh:
    apiserver:
      replicas: 2
      service:
        annotations:
          service.beta.kubernetes.io/aws-load-balancer-internal: "true"
          service.beta.kubernetes.io/aws-load-balancer-type: nlb
          external-dns.alpha.kubernetes.io/hostname: cilium.mercury-dev01.mer-dev02.internal.ABC.cloud
        type: LoadBalancer
      tls:
        auto:
          enabled: false
        server:
          extraDnsNames:
            - cilium.mercury-dev01.mer-dev02.internal.ABC.cloud
          cert: "ABC"
          key: "ABC"
        # -- base64 encoded PEM values for the clustermesh-apiserver admin certificate and private key.
        # Used if 'auto' is not enabled.
        admin:
          cert: "ABC"
          key: "ABC"
        # -- base64 encoded PEM values for the clustermesh-apiserver remote cluster certificate and private key.
        # Used if 'auto' is not enabled.
        remote:
          cert: "ABC"
          key: "ABC"
        client:
          cert: "ABC"
          key: "ABC"
    config:
      clusters:
      - address: cilium.mercury-dev01.mer-dev01.internal.ABC.cloud
        name: mer-dev01-mercury-dev01
        port: 2379
      enabled: true
    useAPIServer: true
  cni:
    binPath: /var/lib/cni/bin
    chainingMode: portmap
    confPath: /var/run/multus/cni/net.d
    # # Needed for Istio
    exclusive: false
  debug:
    enabled: true
  etcd:
    leaseTTL: 30s
  externalWorkloads:
    enabled: true
  hubble: # Enable Hubble and Hubble Relay and avoid UI issues
    enabled: true
    peerService:
      clusterDomain: cluster.local
    relay:
      enabled: true
      tls:
        client:
          cert: "ABC"
          key: "ABC"
    tls:
      auto:
        enabled: false
      server:
        cert: "ABC"
        key: "ABC"
    ui:
      enabled: true
  identityChangeGracePeriod: 0s
  ipam:
    mode: cluster-pool
    operator:
      clusterPoolIPv4MaskSize: "23"
      clusterPoolIPv4PodCIDRList:
      - 100.71.0.0/16
  ipv4:
    enabled: true
  ipv4NativeRoutingCIDR: 100.70.0.0/15
  ipv6:
    enabled: false
  k8s:
    requireIPv4PodCIDR: true
  k8sServiceHost: api.mercury-dev01.mer-dev02.internal.ABC.cloud
  k8sServicePort: 6443
  kubeProxyReplacement: true
  logSystemLoad: true
  prometheus:
    serviceMonitor:
      enabled: false
  rollOutCiliumPods: true
  securityContext:
    privileged: true
  sessionAffinity: true
  tunnelPort: 4789
  tls:
    ca:
      cert: ABC
      key: ABC
  # # Needed for Istio
  socketLB:
    hostNamespaceOnly: true
```
4. Configure Istio with IstioOperator below

Cluster1:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio
  namespace: mercury-istio
spec:
  namespace: mercury-istio
  values:
    profile: openshift
    global:
      istioNamespace: mercury-istio
      meshID: istiomesh
      platform: openshift
      multiCluster:
        clusterName: mer-dev01-mercury-dev01
      network: network1
    cni:
      cniBinDir: /var/lib/cni/bin
      cniConfDir: /etc/cni/multus/net.d
      chained: false
      cniConfFileName: "istio-cni.conf"
      logLevel: info
      provider: "multus"
      excludeNamespaces:
        - kube-system
        - mercury-istio
    pilot:
      cni:
        enabled: true
        provider: "multus"
  components:
    cni:
      enabled: true
      namespace: kube-system
```

Cluster2:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio
  namespace: mercury-istio
spec:
  namespace: mercury-istio
  values:
    profile: openshift
    global:
      istioNamespace: mercury-istio
      meshID: istiomesh
      platform: openshift
      multiCluster:
        clusterName: mer-dev02-mercury-dev01
      network: network1
    cni:
      cniBinDir: /var/lib/cni/bin
      cniConfDir: /etc/cni/multus/net.d
      chained: false
      cniConfFileName: "istio-cni.conf"
      logLevel: info
      provider: "multus"
      excludeNamespaces:
        - kube-system
        - mercury-istio
    pilot:
      cni:
        enabled: true
        provider: "multus"
  components:
    cni:
      enabled: true
      namespace: kube-system
```

### Cilium Version

cilium-cli: v0.16.13 compiled with go1.22.5 on darwin/arm64
cilium image (default): v1.15.6
cilium image (stable): v1.15.7
cilium image (running): 1.15.1

### Kernel Version

Linux ip-10-214-108-135 6.5.5-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Sun Sep 24 15:52:44 UTC 2023 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.30.3
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.1-3474+5c56cc35e3edc1-dirty

### Regression

No

### Sysdump

File too big

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by squeed** [comment]
In your example, you mention that the second service is not global. Can you confirm that, as [per the docs](https://docs.cilium.io/en/latest/network/clustermesh/services/#load-balancing-service-discovery), you've set `service.cilium.io/global: "true"` and `service.cilium.io/shared: "false"`? Without these annotations, service IPs do not work cross-cluster.

Separately, can you validate with a simpler example without Istio? That might help us get to the bottom of the issue.

**Comment by davtex** [comment]
> In your example, you mention that the second service is not global. Can you confirm that, as [per the docs](https://docs.cilium.io/en/latest/network/clustermesh/services/#load-balancing-service-discovery), you've set `service.cilium.io/global: "true"` and `service.cilium.io/shared: "false"`? Without these annotations, service IPs do not work cross-cluster.
> 
> Separately, can you validate with a simpler example without Istio? That might help us get to the bottom of the issue.

I am not using these annotations as I am not interested in Cilium loadbalancing this service. This is Istio's job in my setup. Istio works quite differently and only needs Cilium for cross-cluster inter-pod connectivity.
I did try adding these annotations when I was getting desperate, but it made no effect.

Clustermesh is working as stated in "Use case 1".

**Comment by squeed** [comment]
Aha, right; Istio is doing the load balancing, oops. This is very odd indeed. Is Istio running as a classic per-pod sidecar, or a HostNetwork daemon on the node?

Either way, this is a potentially very complex issue. A sysdump -- shared privately if necessary -- is probably required to diagnose this.

**Comment by davtex** [comment]
@squeed  Istio is running with sidecars, no Ambient or anything fancy.

I have restarted all cilium pods, used curl to call remote pod with success and then used curl to call service without success. I had to remove cilium-bugtool files to lower the size. Kept cilium-bugtool files from source node and target node.

[cluster1.zip](https://github.com/user-attachments/files/16397403/cluster1.zip)
[cluster2.zip](https://github.com/user-attachments/files/16397404/cluster2.zip)

**Comment by youngnick** [comment]
I had a look into the sysdump today, but I can't see any obvious causes. The fact that things work when it's only Clustermesh but don't once we get into service routing that combines Clustermesh and Istio's service discovery points at some interaction between the two types of service discovery to me.

Do the logs of the sidecar proxies indicate anything? Is the connection failing to establish with timeouts (which would indicate packets being dropped somewhere), or are there more specific errors in the proxy logs (which would indicate that the routing is working, but something about the lookups is failing)?

I'd definitely like to echo what @squeed said here, that this is a _very_ complex deployment.

Sadly, I think there's a limited amount we're going to be able to do as a result, aside from suggest places to look for problems.
